sdk/src/types/block/output/delegation.rs

```diff
@@ -355,6 +357,53 @@ impl DelegationOutput {
             .locked_address(self.address(), context.essence.creation_slot())
             .unlock(unlock, inputs, context)
     }
+
+    // Transition, just without full ValidationContext.
+    pub(crate) fn transition_inner(current_state: &Self, next_state: &Self) -> Result<(), StateTransitionError> {
+        if !(current_state.delegation_id.is_null() && !next_state.delegation_id().is_null()) {
+            return Err(StateTransitionError::NonDelayedClaimingTransition);
+        }
+
+        if current_state.delegated_amount != next_state.delegated_amount
+            || current_state.start_epoch != next_state.start_epoch
+            || current_state.validator_id != next_state.validator_id
+        {
+            return Err(StateTransitionError::MutatedImmutableField);
+        }
+
```

Comment:
What about the end epoch rules?

Reply:
https://github.com/iotaledger/iota-sdk/issues/1313